src/s2client-fsharp/ProtobufConnection.fs

```diff
@@ -1,256 +1,168 @@
+            getConnectedSocket()
+
+        let receiveBuf = System.ArraySegment(Array.zeroCreate (1024*1024))
+        let sendBuf = System.ArraySegment(Array.zeroCreate (1024*1024))
+
+        let writeMessage (client:ClientWebSocket) (req:SC2APIProtocol.Request) = async {
+            try
+                use co = new Google.Protobuf.CodedOutputStream(sendBuf.Array)
+                req.WriteTo(co)
+                let written = int co.Position
+                let send = System.ArraySegment(sendBuf.Array, 0, written)
+                do! client.SendAsync(send, WebSocketMessageType.Binary, true, tok) |> Async.AwaitTask
+                return Ok ()
+            with
+            |ex -> return ex.Message |> FailedToSendMessage |> Error
+        }
+
+        let readMessage (client:ClientWebSocket) = async {
+            let rec innerLoop curPos = async {
+                let left = sendBuf.Array.Length - curPos
+                if left <= 0 then
+                    return Error SendMessageBufferTooSmall
+                else
+                    try
+                        let segment = System.ArraySegment(receiveBuf.Array, curPos, left)
+                        let! result = client.ReceiveAsync(segment, tok) |> Async.AwaitTask
+                        match result.MessageType, result.EndOfMessage with
+                        |WebSocketMessageType.Binary, false ->
+                            return! innerLoop (curPos + result.Count)
+                        |WebSocketMessageType.Binary, true -> return Ok (curPos + result.Count)
+                        |_ -> return Error ExpectedBinaryResponse
+                    with
+                    |ex -> return ex.Message |> FailedToReceiveMessage |> Error
+            }
+
+            let parseFrom finalPos =
+                SC2APIProtocol.Response.Parser.ParseFrom(new System.IO.MemoryStream(receiveBuf.Array, 0, finalPos))
+
+            return!
+                innerLoop 0
+                |> Result.mapAsyncInput parseFrom
+        }
+
+        let getAgent client = 
+            let getResponse() =
+                readMessage client
+            MailboxProcessor.Start (fun inbox ->
+                let rec messageLoop() = async{
+                    let! msg = inbox.Receive()
+
+                    match msg with
+                    |SendRequest (req, replyChannel) ->
+                        let! resp =
+                            writeMessage client req
+                            |> Result.bindAsync getResponse
+
+                        replyChannel.Reply resp
+
+                    return! messageLoop()
+                }
+                messageLoop()
+            )
+
+        let postSendRequest req (agent:MailboxProcessor<ClientRequest>) =
+            agent.PostAndAsyncReply (fun reply -> SendRequest (req, reply))
+
+        let sendRequest req = async{
+            return!
+                connectedSocket
+                |> Result.map getAgent
```

Comment:
So you create a new agent every time? Are you sure that is what we want?

Reply:
Good catch! this is not what we want. Will fix it.